### PR TITLE
WT-13166 Instrument __wt_rwlock_* calls to be TSan compatible

### PR DESCRIPTION
**Pull request by ajmorton**
Add dummy atomic reads/writes in our read/write lock functions that use `__ATOMIC_ACQUIRE` and `__ATOMIC_RELEASE`. Our existing locks are safe, this additional information lets TSan understand that and prevents false positive data races for variables protected by locks.

### CODE REVIEW

File: src/support/mtx_rw.c [hunk 1]
```diff
@@ -148,7 +148,14 @@ __wt_try_readlock(WT_SESSION_IMPL *session, WT_RWLOCK *l)
         return (__wt_set_return(session, EBUSY));
 
     /* We rely on this atomic operation to provide a barrier. */
-    return (__wt_atomic_casv64(&l->u.v, old.u.v, new.u.v) ? 0 : EBUSY);
+    WT_RET(__wt_atomic_casv64(&l->u.v, old.u.v, new.u.v) ? 0 : EBUSY);
+
+#ifdef TSAN_BUILD
+    /* Perform a dummy read to inform TSan this function does in fact have acquire semantics. */
+    (void)__atomic_load_n(&l->tsan_sync, __ATOMIC_ACQUIRE);
```
Comment (ajmorton) [inline src/support/mtx_rw.c:155]: I was a bit worried that this void cast would result in reads being optimised out, but using godbolt and looking at test results on the head of the WT-13154 branch these instructions are emitted in compiled code. I'll do some heavy testing to further support this theory.

File: src/include/mutex.h [hunk 2]
```diff
@@ -58,6 +58,15 @@ struct __wt_rwlock { /* Read/write lock */
 
     WT_CONDVAR *cond_readers; /* Blocking readers */
     WT_CONDVAR *cond_writers; /* Blocking writers */
+
+#ifdef TSAN_BUILD
+    /*
+     * Our read/write locks provide thread safety but TSan reports false data races for fields
+     * protected by these locks. To help TSan we perform dummy atomic reads and writes to this field
```
Comment (luke-pearson) [inline src/include/mutex.h:65]: I would clarify the reason TSan reports those races, which is that it doesn't support our barrier implementations which our rw_lock implementation depends on.

File: src/support/mtx_rw.c (hunk 1 above)
Comment (luke-pearson) [reply src/support/mtx_rw.c:155]: I think we shouldn't need any additional semantics after a CAS, have you tested without this one?

File: src/support/mtx_rw.c [hunk 3]
```diff
@@ -339,7 +356,14 @@ __wt_try_writelock(WT_SESSION_IMPL *session, WT_RWLOCK *l)
      */
     new.u.v = old.u.v;
     new.u.s.next++;
-    return (__wt_atomic_casv64(&l->u.v, old.u.v, new.u.v) ? 0 : EBUSY);
+    WT_RET(__wt_atomic_casv64(&l->u.v, old.u.v, new.u.v) ? 0 : EBUSY);
+
+#ifdef TSAN_BUILD
+    /* Perform a dummy write to inform TSan this function does in fact have release semantics. */
```
Comment (luke-pearson) [inline src/support/mtx_rw.c:362]: Again I'm not sure about this one.

File: src/support/mtx_rw.c [hunk 4]
```diff
@@ -454,6 +483,11 @@ __wt_writeunlock(WT_SESSION_IMPL *session, WT_RWLOCK *l)
 {
     WT_RWLOCK new, old;
 
+#ifdef TSAN_BUILD
+    /* Perform a dummy write to inform TSan this function does in fact have release semantics. */
```
Comment (luke-pearson) [inline src/support/mtx_rw.c:487]: I think if anything the release should go at the end of the function, though I am not convinced we need this either.

File: src/support/mtx_rw.c (hunk 1 above)
Comment (ajmorton) [reply src/support/mtx_rw.c:155]: I agree it shouldn't, but local testing seems to require them. If we're adding this reads to some functions I think it's useful to do so consistently, even if only for documentation purposes.

File: src/include/mutex.h (hunk 2 above)
Comment (ajmorton) [reply src/include/mutex.h]: Updated in [fccde73](https://github.com/wiredtiger/wiredtiger/pull/10751/commits/fccde7399554b76646298dcbccfe11a76a8e4555), along with a fixed typo

**Review by luke-pearson — APPROVED** [review]
LGTM.

**Pull request merged**